botamusique: init at unstable-2021-01-02

infinisil opened the pull request:
###### Motivation for this change
This packages [botamusique](https://github.com/azlux/botamusique), a bot to play YouTube, radio, local music and more on Mumble.

This was a bit tricky to package. Some notes on that:
- pymumble is updated to the latest version. However I had to adjust it slightly for pycryptodome compatibility (which is what nixpkgs uses for pycrypto), see https://github.com/azlux/pymumble/pull/99. I have included this as a patch for now
- botamusique isn't a proper python package, it doesn't have a setup.py, see https://github.com/azlux/botamusique/issues/113. So I had to manually wrap its entrypoint script with a python that has all the dependencies, and then create a `botamusique` binary alias for it.
- botamusique has a web frontend, which uses webpack to generate static artifacts. Webpack is based on nodejs, so I used node2nix to generate the dependencies and build the artifacts, mostly as shown [here](https://github.com/svanderburg/node2nix#using-the-nodejs-environment-in-other-nix-derivations). This is only needed at build time.
- For the web frontend, it also uses jinja to instantiate html templates for different languages. So this is also done at build time.
- I made two Nix-specific patches so that it works a little better.
- The most recent stable release has some problems with template path resolving, which is only fixed in master. So I'm using the unstable version now.
- Finally, I added an `updateScript` so that the package can be updated easily in the future.

###### Things done

- [x] Verified that it works with a file `configuration.ini` containing
  ```ini
  [server]
  host = <your-mumble-server>

  [bot]
  volume = 1
  tmp_folder = /tmp/botamusique/

  [webinterface]
  enabled = True
  ```
  Then running
  ```
  $ botamusique --config configuration.ini
  ```
  - [x] Commands work, e.g. `!help` and `!url <youtube-url>`
  - [x] Web interface works, can upload, shuffle, etc. from it
- [x] Verified that the update script works

infinisil committed:
python3Packages.pymumble: 0.3.1 -> 1.6

infinisil committed:
pythonPackages.pyradios: init at 0.0.22

infinisil committed:
botamusique: init at unstable-2021-03-13

SuperSandro2000 commented on pkgs/development/python-modules/pymumble/default.nix:
```diff
@@ -1,31 +1,48 @@
     repo = "pymumble";
-    rev = "1dd6d6d4df2fdef33202f17e2acf3ba9678a5737";
-    sha256 = "1r1sch8xrpbzffsb72lhp5xjr3ac3xb599n44vsfmaam3xklz6vz";
+    rev = "1.5";
```
```suggestion
    rev = version;
```

SuperSandro2000 commented on pkgs/development/python-modules/pymumble/default.nix:
```diff
@@ -1,31 +1,48 @@
+  ];
+
+  postPatch = ''
+    sed -i 's/\(.*\)==.*/\1/' requirements.txt
```
This relaxes the version requirements, right?

Infinisil replied:
Indeed, transforms `foo=1.0` into just `foo`

SuperSandro2000 replied:
I think we might want to comment that.

SuperSandro2000 commented on pkgs/development/python-modules/pymumble/default.nix:
```diff
@@ -1,31 +1,48 @@
 
   propagatedBuildInputs = [ opuslib protobuf ];
 
+  checkInputs = [ pytest pycrypto ];
```
Can we use pytestCheckHook here?

Infinisil replied:
Works without it, so I don't think it's needed

SuperSandro2000 replied:
It probably uses setuptoolsInstallCheck which is deprecated and eventually will be removed.

Infinisil replied:
Ah sounds good, will switch to that then.

SuperSandro2000 commented on pkgs/development/python-modules/pymumble/default.nix:
```diff
@@ -1,31 +1,48 @@
   meta = with lib; {
     description = "Python 3 version of pymumble, Mumble library used for multiple uses like making mumble bot.";
     homepage = "https://github.com/azlux/pymumble";
     license = licenses.gpl3;
```
```suggestion
    license = licenses.gpl3Only;
```

SuperSandro2000 commented on pkgs/tools/audio/botamusique/default.nix:
```diff
@@ -0,0 +1,136 @@
+{ pkgs,
+  lib,
+  stdenv,
+  fetchFromGitHub,
+  python3Packages,
+  coreutils,
+  nix-prefetch-git,
+  jq,
+  nodePackages
+}:
```
```suggestion
{ lib
, stdenv
, fetchFromGitHub
, python3Packages
, coreutils
, nix-prefetch-git
, jq
, nodePackages
}:
```
Generally pkgs shouldn't be used as an input.

Infinisil replied:
`pkgs` is taken by `node-env.nix`, so can't get around it unfortunately

SuperSandro2000 commented on pkgs/tools/audio/botamusique/default.nix:
```diff
@@ -0,0 +1,136 @@
+  nodeEnv = import ../../../development/node-packages/node-env.nix {
+    inherit (pkgs) stdenv lib python2 runCommand writeTextFile;
+    inherit pkgs nodejs;
+    libtool = if pkgs.stdenv.isDarwin then pkgs.darwin.cctools else null;
```
```suggestion
    libtool = lib.optional pkgs.stdenv.isDarwin cctools;
```

I am not sure if this is some node2nix function but generally overwriting packages with null should be avoided.

Infinisil replied:
Yeah these `nodeEnv` and `node-packages.nix` statements are generated by node2nix. But no `lib.optional` won't work here (for one it returns an array, and for another the argument `libtool` isn't optional).

SuperSandro2000 commented on pkgs/tools/audio/botamusique/default.nix:
```diff
@@ -0,0 +1,136 @@
+  })).nodeDependencies;
+
+  # Python needed to instantiate the html templates
+  buildPython = python3Packages.python.withPackages (p: [ p.jinja2 ]);
```
```suggestion
  buildPython = python3Packages.python.withPackages (ps: [ ps.jinja2 ]);
```

thelegy reviewed (approved):
lgtm, but i cant say much about the node part. also built it and tried a couple of things out, and everything i tested worked fine.

SuperSandro2000 commented:
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 112972` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
7 packages built:
</summary>
<ul>
<li>botamusique</li>
<li>python37Packages.pymumble</li>
<li>python37Packages.pyradios</li>
<li>python38Packages.pymumble</li>
<li>python38Packages.pyradios</li>
<li>python39Packages.pymumble</li>
<li>python39Packages.pyradios</li>
</ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

<details>
botamusique:
<details>


`buildPhase` should probably contain `runHook preBuild` and `runHook postBuild`.

Near pkgs/tools/audio/botamusique/default.nix:86:3:

```
   |
86 |   buildPhase = ''
   |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md
`installPhase` should probably contain `runHook preInstall` and `runHook postInstall`.

Near pkgs/tools/audio/botamusique/default.nix:100:3:

```
    |
100 |   installPhase = ''
    |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md
</details>
python37Packages.pyradios:
<details>

Zero tests run by pytest got detected: 'Ran 0 tests in 0.000s'
You have the following options to fix this:
  - Tell pytest(CheckHook) where to find the tests included in the package
  - Check if the GitHub Repo contains tests but they are not shipped with Pypi. If so please switch to `fetchFromGitHub`.
  - If the Packages does not contain any tests add `'doCheck = false;'` and a `pythonImportsCheck`.


Add a `checkPhase` for tests, or at least `pythonImportsCheck`.

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/python-include-tests.md
</details>
python38Packages.pyradios:
<details>

Zero tests run by pytest got detected: 'Ran 0 tests in 0.000s'
You have the following options to fix this:
  - Tell pytest(CheckHook) where to find the tests included in the package
  - Check if the GitHub Repo contains tests but they are not shipped with Pypi. If so please switch to `fetchFromGitHub`.
  - If the Packages does not contain any tests add `'doCheck = false;'` and a `pythonImportsCheck`.


Add a `checkPhase` for tests, or at least `pythonImportsCheck`.

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/python-include-tests.md
</details>
python39Packages.pyradios:
<details>

Zero tests run by pytest got detected: 'Ran 0 tests in 0.000s'
You have the following options to fix this:
  - Tell pytest(CheckHook) where to find the tests included in the package
  - Check if the GitHub Repo contains tests but they are not shipped with Pypi. If so please switch to `fetchFromGitHub`.
  - If the Packages does not contain any tests add `'doCheck = false;'` and a `pythonImportsCheck`.


Add a `checkPhase` for tests, or at least `pythonImportsCheck`.

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/python-include-tests.md
</details>
</details>

SuperSandro2000 commented:
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 112972` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
7 packages built:
</summary>
<ul>
<li>botamusique</li>
<li>python37Packages.pymumble</li>
<li>python37Packages.pyradios</li>
<li>python38Packages.pymumble</li>
<li>python38Packages.pyradios</li>
<li>python39Packages.pymumble</li>
<li>python39Packages.pyradios</li>
</ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

<details>
botamusique:
<details>


`buildPhase` should probably contain `runHook preBuild` and `runHook postBuild`.

Near pkgs/tools/audio/botamusique/default.nix:86:3:

```
   |
86 |   buildPhase = ''
   |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md
`installPhase` should probably contain `runHook preInstall` and `runHook postInstall`.

Near pkgs/tools/audio/botamusique/default.nix:100:3:

```
    |
100 |   installPhase = ''
    |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md
</details>
python37Packages.pyradios:
<details>

Zero tests run by pytest got detected: 'Ran 0 tests in 0.000s'
You have the following options to fix this:
  - Tell pytest(CheckHook) where to find the tests included in the package
  - Check if the GitHub Repo contains tests but they are not shipped with Pypi. If so please switch to `fetchFromGitHub`.
  - If the Packages does not contain any tests add `'doCheck = false;'` and a `pythonImportsCheck`.


Add a `checkPhase` for tests, or at least `pythonImportsCheck`.

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/python-include-tests.md
</details>
python38Packages.pyradios:
<details>

Zero tests run by pytest got detected: 'Ran 0 tests in 0.000s'
You have the following options to fix this:
  - Tell pytest(CheckHook) where to find the tests included in the package
  - Check if the GitHub Repo contains tests but they are not shipped with Pypi. If so please switch to `fetchFromGitHub`.
  - If the Packages does not contain any tests add `'doCheck = false;'` and a `pythonImportsCheck`.


Add a `checkPhase` for tests, or at least `pythonImportsCheck`.

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/python-include-tests.md
</details>
python39Packages.pyradios:
<details>

Zero tests run by pytest got detected: 'Ran 0 tests in 0.000s'
You have the following options to fix this:
  - Tell pytest(CheckHook) where to find the tests included in the package
  - Check if the GitHub Repo contains tests but they are not shipped with Pypi. If so please switch to `fetchFromGitHub`.
  - If the Packages does not contain any tests add `'doCheck = false;'` and a `pythonImportsCheck`.


Add a `checkPhase` for tests, or at least `pythonImportsCheck`.

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/python-include-tests.md
</details>
</details>

thelegy reviewed (commented):
Can you please address the points from @SuperSandro2000 or comment, why they dont apply.

Infinisil commented:
They're all addressed/implemented already, that's why I marked them as resolved :)

thelegy commented:
Sorry, but I cant find where you addressed a couple of them, could you please point me to there? Especially i have trouble finding your reaction on the following points:

- botamusique: `buildPhase` should probably contain `runHook preBuild` and `runHook postBuild`.
- botamusique: `installPhase` should probably contain `runHook preInstall` and `runHook postInstall.`
- pyradios: Zero tests run by pytest got detected

Infinisil commented:
@thelegy Ah I see, automated messages do have the tendency to be ignored. I addressed it now

I also rebased this PR ontop of latest master, updated to a newer pymumble release and the latest botamusique commit. Then made sure to test everything again. I'll merge after checks succeed
